### ISSUE TITLE
For mixed-precision training, does FSDP2 also need `amp.grad_scaler.GradScaler` ? or is FSDP2 already handled?

### ISSUE THREAD

**Issue by EquationWalker** [pytorch/torchtitan#1735]
In mixed-precision training of DDP, `amp.grad_scaler.GradScaler`  is needed to dynamically scale the loss. I see that torchtitan do not use it to scale loss in FSDP2, so my question is does FSDP2 also need `amp.grad_scaler.GradScaler` ? or is FSDP2 already handled?

**Comment by tianyu-l** [comment]
IIUC FSDP2 scales the gradient like `sum(gradients from all FSDP ranks) / number of FSDP ranks`. For the dividing factor, you can manually adjust use `set_gradient_divide_factor`, e.g. https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/llama4/infra/parallelize.py#L362

@weifengpy can share more.

update: oh what I said may not be relevant

**Comment by weifengpy** [comment]
this is the same quesiton to pytorch issue https://github.com/pytorch/pytorch/issues/163519

titan shows the best practice of fully_shard(model_fp_32, mp_policy), where model_fp_32 is fp32 model.parameters()
* inside forward/backward, param dtype (can be fp16) and reduce dtype (usually fp32) are decided by mp_policy
* outside forwad/backward, model.parameters() are fp32. thus no need for grad scaler

~~the only situation is that need grad scaler is fully_shard(model_fp_16). model.parameters() are fp16 outside of forward/backward~~ (editted: we need grad scaler for mp_policy.param_dtype=fp16)

wonder which DDP mixed precision are you referring to? is it torch.amp.autocast or DDP(mp_policy)? if DDP(mp_policy), it's the same rule to FSDP

**Comment by EquationWalker** [comment]
> this is the same quesiton to pytorch issue [pytorch/pytorch#163519](https://github.com/pytorch/pytorch/issues/163519)
> 
> titan shows the best practice of fully_shard(model_fp_32, mp_policy), where model_fp_32 is fp32 model.parameters()
> 
> * inside forward/backward, param dtype (can be fp16) and reduce dtype (usually fp32) are decided by mp_policy
> * outside forwad/backward, model.parameters() are fp32. thus no need for grad scaler
> 
> the only situation is that need grad scaler is fully_shard(model_fp_16). model.parameters() are fp16 outside of forward/backward
> 
> wonder which DDP mixed precision are you referring to? is it torch.amp.autocast or DDP(mp_policy)? if DDP(mp_policy), it's the same rule to FSDP

Thanks for your reply！The reason I have this question is because I see [torchtitan/docs/fsdp.md](https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md?plain=1#L83) that `amp.grad_scaler.GradScaler` is similar to `ShardedGradScaler` in FSDP1. I just checked the [FSDP2 source code](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py#L728) and it looks like it handles this internally, so you can check it.

**Comment by EquationWalker** [comment]
> IIUC FSDP2 scales the gradient like `sum(gradients from all FSDP ranks) / number of FSDP ranks`. For the dividing factor, you can manually adjust use `set_gradient_divide_factor`, e.g. https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/llama4/infra/parallelize.py#L362
> 
> [@weifengpy](https://github.com/weifengpy) can share more.
> 
> update: oh what I said may not be relevant

Thanks for your reply!

**Comment by EquationWalker** [comment]
> > this is the same quesiton to pytorch issue [pytorch/pytorch#163519](https://github.com/pytorch/pytorch/issues/163519)
> > titan shows the best practice of fully_shard(model_fp_32, mp_policy), where model_fp_32 is fp32 model.parameters()
> > 
> > * inside forward/backward, param dtype (can be fp16) and reduce dtype (usually fp32) are decided by mp_policy
> > * outside forwad/backward, model.parameters() are fp32. thus no need for grad scaler
> > 
> > the only situation is that need grad scaler is fully_shard(model_fp_16). model.parameters() are fp16 outside of forward/backward
> > wonder which DDP mixed precision are you referring to? is it torch.amp.autocast or DDP(mp_policy)? if DDP(mp_policy), it's the same rule to FSDP
> 
> Thanks for your reply！The reason I have this question is because I see [torchtitan/docs/fsdp.md](https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md?plain=1#L83) that `amp.grad_scaler.GradScaler` is similar to `ShardedGradScaler` in FSDP1. I just checked the [FSDP2 source code](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py#L728) and it looks like it handles this internally, so you can check it.

As [pytorch/issues/163519#issuecomment-3322148922](https://github.com/pytorch/pytorch/issues/163519#issuecomment-3322148922) says, the overflow at grad reduction and the overflow for initial loss multiplication are two different things.

**Comment by EquationWalker** [comment]
> > > this is the same quesiton to pytorch issue [pytorch/pytorch#163519](https://github.com/pytorch/pytorch/issues/163519)
> > > titan shows the best practice of fully_shard(model_fp_32, mp_policy), where model_fp_32 is fp32 model.parameters()
> > > 
> > > * inside forward/backward, param dtype (can be fp16) and reduce dtype (usually fp32) are decided by mp_policy
> > > * outside forwad/backward, model.parameters() are fp32. thus no need for grad scaler
> > > 
> > > the only situation is that need grad scaler is fully_shard(model_fp_16). model.parameters() are fp16 outside of forward/backward
> > > wonder which DDP mixed precision are you referring to? is it torch.amp.autocast or DDP(mp_policy)? if DDP(mp_policy), it's the same rule to FSDP
> > 
> > 
> > Thanks for your reply！The reason I have this question is because I see [torchtitan/docs/fsdp.md](https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md?plain=1#L83) that `amp.grad_scaler.GradScaler` is similar to `ShardedGradScaler` in FSDP1. I just checked the [FSDP2 source code](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py#L728) and it looks like it handles this internally, so you can check it.
> 
> As [pytorch/issues/163519#issuecomment-3322148922](https://github.com/pytorch/pytorch/issues/163519#issuecomment-3322148922) says, the overflow at grad reduction and the overflow for initial loss multiplication are two different things.

As [Gradient scaler for DTensor #132816](https://github.com/pytorch/pytorch/pull/132816) said, perhaps FSDP2 requires scaler for fp16 training? @weifengpy

**Comment by weifengpy** [comment]
> As [Gradient scaler for DTensor #132816](https://github.com/pytorch/pytorch/pull/132816) said, perhaps FSDP2 requires scaler for fp16 training? @weifengpy 

 thanks for the link! I know this PR when reply. for fully_shard(model_fp32, mp_policy), opt.step will be in fp32 entirely. thus no need to grad scaler

**Comment by EquationWalker** [comment]
> > As [Gradient scaler for DTensor #132816](https://github.com/pytorch/pytorch/pull/132816) said, perhaps FSDP2 requires scaler for fp16 training? [@weifengpy](https://github.com/weifengpy)
> 
> thanks for the link! I know this PR when reply. for fully_shard(model_fp32, mp_policy), opt.step will be in fp32 entirely. thus no need to grad scaler

When `mp_policy.param_dtype=fp16` and `model_fp32`, the gradient is computed in fp16 in the backward (`FSDPParam._unshared_param.grad`). Underflow may occur during the backward fp16 computation, even if the gradient is subsequently converted to fp32 (`FSDPParam.shared_param.grad`). Underflow occurs in the backward fp16 computation rather than in the step of the optimizer.

**Comment by weifengpy** [comment]
> When `mp_policy.param_dtype=fp16` and `model_fp32`, the gradient is computed in fp16 in the backward (`FSDPParam._unshared_param.grad`). Underflow may occur during the backward fp16 computation, even if the gradient is subsequently converted to fp32 (`FSDPParam.shared_param.grad`). Underflow occurs in the backward fp16 computation rather than in the step of the optimizer.

@EquationWalker you're right! sorry I didn't understand GradScaler before. we need GradScaler for mp_policy.param_dtype=fp16

for people who googled the issue, because fsdp2 use DTensor, it's plain torch.amp.grad_scaler  https://github.com/pytorch/pytorch/pull/132816

```
from torch.amp.grad_scaler import GradScaler
scaler = GradScaler(init_scale=2.0, enabled=True)
opt = torch.optim.Adam(model.parameters(), lr=1e-2)
for _ in range(10)
    loss = fsdp2_model(input).sum()
    scaler.scale(loss).backward()
    scaler.step(opt)
    scaler.update()
    opt.zero_grad()
```
